src/sgame/lua/Interpreter.cpp

```diff
@@ -0,0 +1,239 @@
+	for ( i = 1; i <= n; ++i )
+	{
+		const char* s;
+		lua_pushvalue( L, -1 ); /* function to be called */
+		lua_pushvalue( L, i );  /* value to print */
+		lua_call( L, 1, 1 );
+		s = lua_tostring( L, -1 ); /* get result */
+		if ( s == NULL ) return luaL_error( L, "'tostring' must return a string to 'print'" );
+		if ( i > 1 ) output += "\t";
+		output += s;
+		lua_pop( L, 1 ); /* pop result */
+	}
+	Log::Notice( output );
+	return 0;
+}
+
+static int LoadUnvModule( lua_State* L )
+{
+	const char* file = luaL_checkstring( L, 1 );
+	if ( !file )
+	{
+		Log::Warn( "invalid call to LoadUnvModule: must pass in a string." );
+		return 0;
+	}
+	if ( !LoadScript( file ) )
+	{
+		Log::Warn( "error loading script %s", file );
+		return 0;
+	}
+	return 1;
+}
+
+// Remove searchers that search the filesystems so we can install our own searcher.
+// Keep the searcher at index 1 because it just looks for cached modules that have
+// already been loaded.
+const char INITIALIZE_SEARCHERS[] = R"(
+table.remove(package.searchers, 2)
+table.remove(package.searchers, 3)
+)";
+
+static void OverrideGlobalLuaFunctions()
+{
+	lua_getglobal( L, "_G" );
+
+	lua_pushcfunction( L, LuaPrint );
+	lua_setfield( L, -2, "print" );
+
+	if ( !ExecCode( INITIALIZE_SEARCHERS, __func__ ) )
+	{
+		Sys::Drop( "Can't register searcher: failed to run initializer code" );
+	}
+	// Add the package loader to the package.loaders table.
+	lua_getglobal( L, "package" );
+	lua_getfield( L, -1, "searchers" );
+
+	if ( lua_isnil( L, -1 ) )
+	{
+		Sys::Drop( "Can't register searcher: package.loaders table does not exist." );
+	}
+
+	lua_pushcfunction( L, LoadUnvModule );
+	lua_rawseti( L, -2, 2 );
+	lua_pop( L, 2 );
+}
+
+void Initialize()
+{
+	if ( L ) return;
+
+	L = luaL_newstate();
+	luaL_openlibs( L );
+	OverrideGlobalLuaFunctions();
+	BG_InitializeLuaConstants( L );
+}
+
+void Shutdown()
+{
+	lua_close( L );
+	L = nullptr;
+}
+
+lua_State* State()
+{
+	return L;
+}
+
+class LuaCommand : Cmd::StaticCmd
+{
+   public:
+	LuaCommand() : Cmd::StaticCmd( "lua", Cmd::SGAME_VM, "exec server side lua" ) {}
+
+	void Run( const Cmd::Args& args ) const
+	{
+		auto usage = [ & ]()
+		{ PrintUsage( args, "[-f filename ] | <lua code>", "exec server side lua" ); };
+		if ( args.Argc() < 2 )
+		{
+			usage();
+			return;
+		}
+		const std::string& first = args.Argv( 1 );
+		if ( first == "-f" )
+		{
+			if ( args.Argc() < 3 )
+			{
+				Log::Warn( "Must pass in filename when using the '-f' flag." );
+				usage();
+				return;
+			}
+			// These are arguments that are passed to the script and
+			// accessible to the script in the `arg` array.
+			if ( args.Argc() > 3 )
+			{
+				lua_newtable( L );
+				for ( int i = 3; i < args.Argc(); ++i )
+				{
+					lua_pushstring( L, args.Argv( i ).c_str() );
+					lua_rawseti( L, -2, i - 2 );  // lua arrays start at 1
+				}
+				lua_setglobal( L, "arg" );
+			}
+			ExecScript( args.Argv( 2 ) );
+			return;
+		}
+		ExecCode( first, "/lua" );
```

Comment:
heh, I'm rediscovering a lot from 2 years ago. Turns I had originally done this but ConcatArgs fucks up quoting, so we can't use it.

ie, `lua print("hello")` becomes `lua print(hello)`, which obviously is wrong since hello gets turned from a string to a variable.

Reply:
it's really weird to have to escape quotes without quotes

Reply:
I'm just saying ConcatArgs makes more things work without breaking anything, so I don't see a reason not to do it.

Reply:
Done.